source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,591 @@
+                for (unsigned int point=0; point<n_fs_face_q_points; ++point)
+                  {
+                    // Get the gravity vector to compute the outward direction of displacement
+                    Tensor<1,dim> direction = -(this->get_gravity_model().gravity_vector(fs_fe_face_values.quadrature_point(point)));
+                    // Normalize direction vector
+                    if (direction.norm() > 0.0)
+                      direction *= 1./direction.norm();
+                    // TODO this is only correct for box geometries
+                    else
+                      direction[dim-1] = 1.;
+
+                    // Compute the total displacement in the gravity direction,
+                    // i.e. the initial topography + any additional mesh displacement.
+                    const double displacement = direction * (displacement_values[point] + initial_topography_values[point]);
+
+                    // To project onto the tangent space of the surface,
+                    // we define the projection P:= I- n x n,
+                    // with I the unit tensor and n the unit normal to the surface.
+                    // The surface gradient then is P times the usual gradient of the shape functions.
+                    const Tensor<2, dim, double> projection = unit_symmetric_tensor<dim>() -
+                                                              outer_product(fs_fe_face_values.normal_vector(point), fs_fe_face_values.normal_vector(point));
+
+                    // Loop over the shape functions
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first != dim-1)
+                          continue;
+
+                        // Assemble the RHS
+                        // RHS = M*H_old
+                        cell_vector(i) += displacement *
+                                          fs_fe_face_values.shape_value (i, point) *
+                                          fs_fe_face_values.JxW(point);
+
+
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          {
+                            // Make sure we only assemble for the y-component
+                            // TODO this is only correct for box geometries
+                            if (mesh_deformation_dof_handler.get_fe().system_to_component_index(j).first != dim-1)
+                              continue;
+                            // Assemble the matrix, for backward first order time discretization:
+                            // Matrix := (M+dt*K) = (M+dt*B^T*kappa*B)
+                            cell_matrix(i,j) +=
+                              (
+                                this->get_timestep() * diffusivity *
```

Comment:
I think this is wrong. If you don't run this function in every time step, then you will end up with a dt that corresponds to one time step, but not to the interval between two solves.

My suggestion would be to run this every time step. That avoids having to think what dt you want to use here, and avoids some difficulty with the run-time parameters

Reply:
I don't understand why this is wrong. We always take the topography of the previous timestep to diffuse to compute the current topography. We do not store the topography for the next time we diffuse it, so why does it matter that the timestep is not the interval between to diffusion solves?

Reply:
Oh, I see, you think of the "every once in a while" thing not as "using a larger time step for the diffusion than for the flow", but really only as "I'm fixing up the boundary a little bit every once in a while". That corresponds to having no diffusion for the majority of time steps, and only occasionally having a little bit of diffusion -- a rain burst in the desert.

But is that then really what you want? If, for example, you only do the diffusion step once every million years, then the amount of diffusion you get per million years is proportional to the length of the time step you happen to be in during that one time step when you hit the next time threshold. That means in particular that if you refine the mesh, your time step goes down, and the amount of diffusion goes down. 

In different words: You're not going to *converge* with this scheme as you refine the mesh because the amount of diffusion in the model depends on the length of the time step.

Reply:
Thanks for the offline talk. I'll remove the `time_between_diffusion` parameter.